rust/otap-dataflow/crates/quiver/src/segment/types.rs

```diff
@@ -0,0 +1,576 @@
+// ─────────────────────────────────────────────────────────────────────────────
+
+/// Metadata for a single stream within a segment.
+///
+/// This information is stored in the segment's **stream directory** and allows
+/// readers to locate and interpret stream data without scanning the entire file.
+/// Each stream contains Arrow IPC data for a specific `(slot, schema)` pairing.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct StreamMetadata {
+    /// Unique identifier for this stream within the segment.
+    pub id: StreamId,
+    /// The slot this stream serves.
+    pub slot_id: SlotId,
+    /// Schema fingerprint for all batches in this stream.
+    pub schema_fingerprint: SchemaFingerprint,
+    /// Byte offset from segment start to the stream's Arrow IPC data.
+    pub byte_offset: u64,
+    /// Total byte length of the stream's Arrow IPC data.
+    pub byte_length: u64,
+    /// Total number of rows across all chunks in this stream.
+    pub row_count: u64,
+    /// Number of chunks in this stream. Each chunk corresponds to one
+    /// Arrow `RecordBatch` written from a [`RecordBundle`](crate::record_bundle::RecordBundle) payload.
+    pub chunk_count: u32,
+}
+
+impl StreamMetadata {
+    /// Creates new stream metadata.
+    #[must_use]
+    pub fn new(
+        id: StreamId,
+        slot_id: SlotId,
+        schema_fingerprint: SchemaFingerprint,
+        byte_offset: u64,
+        byte_length: u64,
+        row_count: u64,
+        chunk_count: u32,
+    ) -> Self {
+        Self {
+            id,
+            slot_id,
+            schema_fingerprint,
+            byte_offset,
+            byte_length,
+            row_count,
+            chunk_count,
+        }
+    }
+
+    /// Returns the stream key for this metadata.
+    #[must_use]
+    pub fn stream_key(&self) -> StreamKey {
+        StreamKey::new(self.slot_id, self.schema_fingerprint)
+    }
+}
+
+// ─────────────────────────────────────────────────────────────────────────────
+// Chunk References
+// ─────────────────────────────────────────────────────────────────────────────
+
+/// Zero-based index of a chunk within a stream.
+///
+/// Each Arrow `RecordBatch` written to a stream becomes a chunk at a
+/// specific index. When a [`RecordBundle`] payload is appended to a stream,
+/// the resulting chunk index is recorded in the batch manifest so the
+/// original bundle can be reconstructed during reads.
+///
+/// [`RecordBundle`]: crate::record_bundle::RecordBundle
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub struct ChunkIndex(u32);
+
+impl ChunkIndex {
+    /// Creates a new chunk index.
+    #[must_use]
+    pub const fn new(raw: u32) -> Self {
+        Self(raw)
+    }
+
+    /// Returns the raw numeric value.
+    #[inline]
+    #[must_use]
+    pub const fn raw(self) -> u32 {
+        self.0
+    }
+
+    /// Returns the next sequential chunk index.
+    #[must_use]
+    pub const fn next(self) -> Self {
+        Self(self.0 + 1)
```

Comment:
Can this overflow for long running stream, maybe  clamp with `saturating_add(1)` ?